File: lib/java/tcollier/Executor.java
package tcollier;

import java.lang.StringBuilder;
import java.util.ArrayList;

class IterationTimer {
  private long durationInMillis;
  private int iterations;

  public IterationTimer(long durationInMillis, int iterations) {
    this.durationInMillis = durationInMillis;
    this.iterations = iterations;
  }

  public void appendJson(StringBuilder builder) {
    builder.append("{\"duration\":");
    builder.append(durationInMillis * 1000);
    builder.append(",\"iterations\":");
    builder.append(iterations);
    builder.append("}");
  }
}

public class Executor<T> {
  private Solution<T> solution;
  private ArrayList<T> input;

  public Executor(Solution<T> solution, ArrayList<T> input) {
    this.solution = solution;
    this.input = input;
  }

  public void run(String[] args) {
    if (args.length >= 1 && args[args.length - 1].equals("--time")) {
      time();
    } else {
      solve();
    }
  }

  private void solve() {
    System.out.println(solution.part1Answer((ArrayList<T>)input.clone()));
    System.out.println(solution.part2Answer((ArrayList<T>)input.clone()));
  }

  private void time() {
    IterationTimer part1 = timeAnswer(1);
    IterationTimer part2 = timeAnswer(2);
    StringBuilder builder = new StringBuilder();
    builder.append("{\"part1\":");
    part1.appendJson(builder);
    builder.append(",\"part2\":");
    part2.appendJson(builder);
    builder.append("}");
    System.out.println(builder.toString());
  }

  private boolean continueTiming(int iterations, long duration) {
    if (iterations < 100) {
      return duration < 30000l;
    } else {
      return duration < 100l;
    }
  }

  private IterationTimer timeAnswer(int part) {
    int i = 0;
    long runningTime = 0;
    for (; continueTiming(i, runningTime); i++) {
      ArrayList<T> inputClone = (ArrayList<T>)input.clone();
      long startTime = System.currentTimeMillis();
      if (part == 1) {
        solution.part1Answer(inputClone);
      } else {
        solution.part2Answer(inputClone);
      }
      runningTime += System.currentTimeMillis() - startTime;
    }
    return new IterationTimer(runningTime, i);
  }
}

File: lib/scala/tcollier/Executor.scala
package tcollier

class IterationTimer(val durationInMillis: Long, val iterations: Int) {
  def asJson(): String = {
    return s"""{"duration":${durationInMillis * 1000},"iterations":${iterations}}"""
  }
}

class Executor[T](solution: Solution[T] , input: Array[T]) {
  def run(args: Array[String]) = {
    if (args.size >= 1 && args(args.size - 1).equals("--time")) {
      time()
    } else {
      solve()
    }
  }

  def solve() = {
    println(solution.part1Answer(input))
    println(solution.part2Answer(input))
  }

  def time() = {
    val part1: IterationTimer = timeAnswer(1)
    val part2: IterationTimer = timeAnswer(2)
    println(s"""{"part1":${part2.asJson()},"part2":${part2.asJson()}}""")
  }

  def continueTiming(iterations: Int, duration: Long): Boolean = {
    if (iterations < 100) {
      return duration < 30000L
    } else {
      return duration < 100L
    }
  }

  def timeAnswer(part: Int): IterationTimer = {
    var i: Int = 0
    var runningTime: Long = 0
    while (continueTiming(i, runningTime)) {
      val List[T] inputClone = (List<T>)input.clone()
      val startTime: Long = System.currentTimeMillis()
      if (part == 1) {
        solution.part1Answer(inputClone)
      } else {
        solution.part2Answer(inputClone)
      }
      runningTime += System.currentTimeMillis() - startTime
      i += 1
    }
    return new IterationTimer(runningTime, i)
  }
}
